log_cache_async.go

```diff
@@ -0,0 +1,405 @@
+}
+
+func (c *LogCacheAsync) runFlusher() {
+	var batch []*Log
+
+	for {
+		syncReq := <-c.state.triggerChan
+
+		// Load the state under lock
+		c.state.Lock()
+		persistedIdx := atomic.LoadUint64(&c.persistentIndex)
+		lastIdx := atomic.LoadUint64(&c.lastIndex)
+
+		for idx := persistedIdx + 1; idx <= lastIdx; idx++ {
+			batch = append(batch, c.state.cache[idx&c.sizeMask])
+		}
+
+		// If we are stopping that means the serial writer is blocked in
+		// DisableAsync waiting for us to be done syncing. Keep hold of the lock
+		// while we sync everything that is unflushed and then stop.
+		if syncReq.stop {
+			lwc := c.doFlush(batch, syncReq.startTime)
+			c.deliverCompletionLocked(lwc)
+
+			// Teardown async state here while we hold the lock to make it simpler to
+			// reason about possible races.
+			c.state.completionCh = nil
+			close(syncReq.doneCh)
+
+			// Release the lock or we'll deadlock this node next time it's elected!
+			c.state.Unlock()
+			return
+		}
+
+		// Not stopping, unlock to allow concurrent writes while we flush this batch
+		// to disk.
+		c.state.Unlock()
+
+		// Might be a no-op if batch is empty
+		lwc := c.doFlush(batch, syncReq.startTime)
+
+		// Note: if the flush failed we might retry it on the next loop. This is
+		// safe assuming that the LogStore is atomic and not left in an invalid
+		// state (which Raft assumes in general already). It might loop and retry
+		// the write of the same logs next time which may fail again or may even
+		// succeed before the leaderloop notices the error and steps down. But
+		// either way it's fine because we don't advance the persistedIndex if it
+		// fails so we'll keep trying to write the same logs at least not leave a
+		// gap. Actually if we do error, even if there is no immediate sync trigger
+		// waiting, the leader will step down and disable async which will mean we
+		// attempt to flush again anyway. If that fails though (in the stop case
+		// above) we won't keep retrying and will just re-report the error.
+
+		// Need a lock to deliver the completion and update persistent index
+		c.state.Lock()
+		c.deliverCompletionLocked(lwc)
+		c.state.Unlock()
+
+		close(syncReq.doneCh)
+
+		// Loop around, if more writes have happened, triggerChan will already have
+		// a syncReq buffered and we'll immediately start flushing again.
+	}
+}
+
+func (c *LogCacheAsync) deliverCompletionLocked(lwc *LogWriteCompletion) {
+	if lwc == nil {
+		return
+	}
+	if lwc.Error == nil {
+		atomic.StoreUint64(&c.persistentIndex, lwc.PersistentIndex)
+	}
+	c.state.completionCh <- *lwc
+}
+
+func (c *LogCacheAsync) doFlush(logs []*Log, start time.Time) *LogWriteCompletion {
+	if len(logs) < 1 {
+		return nil
+	}
+
+	// Write to the underlying store
+	err := c.store.StoreLogs(logs)
+
+	lwc := LogWriteCompletion{
+		PersistentIndex: logs[0].Index - 1,
+		Error:           err,
+		Duration:        time.Since(start),
+	}
+	if err == nil {
+		lwc.PersistentIndex = logs[len(logs)-1].Index
+	}
```

Comment:
Yep, here is bug #3 as I guessed. We are not updating `c.persistentIndex` here so the next flush is going to flush all the same logs again and never "empty" the cache.